{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## \"EvoMPNN\" preparation notebook\n",
    "\n",
    "### Goal of this notebook\n",
    "* This notebook takes in a multiple sequence aligment (MSA) for a protein of interest and outputs residues to design for MPNN\n",
    "* Residues are selected following the [\"EvoMPNN\" paper](https://www.biorxiv.org/content/10.1101/2023.10.03.560713v1.abstract)\n",
    "    * The paper did this by ranking the frequency of the most common amino acid at all positions in the MSA\n",
    "    * Then, the top 30%, 50%, or 70% of the most frequent residues were locked, i.e. not designed, during MPNN\n",
    "    * Empirically, enzymes with 50% or 70% locked residues performed well in that paper, so this notebook focuses on those\n",
    "    * I also made it the case that the residue is only locked if it's residue is the same as the most frequent residue\n",
    "    * If you want other cutoff values, this can be done by changing the cutoff argument when calling cutoff_ranked_resis()\n",
    "* The final cell (bottom of the script) will print the residues to lock in the format that MPNN wants, a space delimted list.\n",
    "\n",
    "### How to use:\n",
    "1. Make an MSA in .a3m format using a tool such as hhblits [(link here)](https://toolkit.tuebingen.mpg.de/tools/hhblits)\n",
    "    - Input your sequence in fasta format\n",
    "    - When the run is finished, go to the Query MSA tab and click Download Full A3M\n",
    "2. Update the variables in the cell directly below this to the .a3m path and the name of your protein \n",
    "    - The protein name should be the same as the first sequence in the output .a3m file\n",
    "    - You can add additional residues to lock (e.g. active site residues) as the resis_to_lock list\n",
    "3. Run all cells (spam shift+enter)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "input_a3m = ''\n",
    "input_protein = ''\n",
    "resis_to_lock = []\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# these all assume that your input protein at the top of your a3m file has no '-' characters!\n",
    "\n",
    "def process_fasta(fasta_file):\n",
    "    with open(fasta_file) as f:\n",
    "        fasta_dict = {}\n",
    "        lines = [l.strip() for l in f.readlines() if l.strip()]\n",
    "        for i in range(len(lines)):\n",
    "            if lines[i][0] == '>':\n",
    "                name = lines[i][1:]\n",
    "                seq = ''\n",
    "                new_index = i + 1\n",
    "                while lines[new_index][0] != '>':\n",
    "                    # seq += lines[new_index] #.strip()\n",
    "                    upper_only = ''\n",
    "                    for chr in lines[new_index]:\n",
    "                        if chr.isupper() or chr == '-':\n",
    "                            upper_only += chr\n",
    "                    seq += upper_only\n",
    "                    new_index += 1\n",
    "                    if new_index == len(lines):\n",
    "                        break\n",
    "                \n",
    "                fasta_dict[name] = seq\n",
    "    \n",
    "    return fasta_dict\n",
    "\n",
    "def normalize_freqs(d):\n",
    "    total_resis = sum([d[i] for i in d if i != '-'])\n",
    "    perc_freqs = {}\n",
    "    \n",
    "    for res, num in d.items():\n",
    "        if res != '-':\n",
    "            perc_freqs[res] = num / total_resis\n",
    "    \n",
    "    max_resi = max(perc_freqs, key=perc_freqs.get)\n",
    "    max_freq = perc_freqs[max_resi]\n",
    "\n",
    "    for resi, freq in perc_freqs.items():\n",
    "        if freq >= max_freq and resi != max_resi:\n",
    "            max_resi += resi\n",
    "            max_freq += freq\n",
    "\n",
    "    return max_resi, max_freq\n",
    "\n",
    "def get_consensus_dict(fasta_dict):\n",
    "    # output: a dict of res_pos keys, (most common aa, aa freq) values\n",
    "    seqs = list(fasta_dict.values())\n",
    "    \n",
    "    # check all seq lens are the same\n",
    "    if len({len(seq) for seq in seqs}) != 1:\n",
    "        return 'not all seqs are the same len! check alignment'\n",
    "    \n",
    "    # num_aas_by_pos will be {resnum: {aa1: num, aa2: num}}\n",
    "    # everything will be 1-indexed to be consistent with protein numbering, not standard python\n",
    "    num_aas_by_pos = {i:{} for i in range(1, len(seqs[0])+1)}\n",
    "    for seq in seqs:\n",
    "        for idx, aa in enumerate(seq):\n",
    "            if aa not in num_aas_by_pos[idx+1]:\n",
    "                num_aas_by_pos[idx+1][aa] = 1\n",
    "            else:\n",
    "                num_aas_by_pos[idx+1][aa] += 1\n",
    "    \n",
    "    # perc_aas_by_pos is {resnum: (one letter res, percent freq)}\n",
    "    # if two or more resis have the same freq, one letter res will be a str of all one letter codes and percent freq will be summed\n",
    "    # skip \"-\" (i.e. gap) characters\n",
    "    perc_aas_by_pos = {}\n",
    "    consensus = ['' for _ in range(len(seqs[0]))]\n",
    "    for resnum, aa_freq in num_aas_by_pos.items():\n",
    "        max_resi, max_freq = normalize_freqs(aa_freq)\n",
    "        perc_aas_by_pos[resnum] = (max_resi, max_freq)\n",
    "        consensus[resnum-1] = max_resi\n",
    "\n",
    "    return perc_aas_by_pos, consensus\n",
    "\n",
    "# goal: rank the conservation of residues and take only the top x% that are conserved.\n",
    "\n",
    "def cutoff_ranked_resis(perc_aas_by_pos, cutoff_perc):\n",
    "    # perc_aas_by_pos is {resnum: (one letter res, percent freq)}\n",
    "\n",
    "    # ordered resis should be a list of (resnum, one letter res, percent freq)\n",
    "    ordered_resis = [(i, *perc_aas_by_pos[i]) for i in range(1, len(perc_aas_by_pos)+1)]\n",
    "    \n",
    "    ordered_resis.sort(key=lambda x:x[2], reverse=True)\n",
    "    cutoff_num = cutoff_perc * len(ordered_resis)\n",
    "    \n",
    "    return ordered_resis[:int(cutoff_num)]\n",
    "\n",
    "def renum_ranked_to_seq(seqname, consensus, seq_dict, ranked_resis, keep_same_as_consensus=False):\n",
    "\n",
    "    seq = seq_dict[seqname]\n",
    "    keep_resis = set([i[0] for i in ranked_resis])\n",
    "\n",
    "    # need to ensure that we are getting the consensus at each position\n",
    "\n",
    "    mapped_seq = []\n",
    "\n",
    "    for idx, seq_aa in enumerate(seq):\n",
    "        assert seq_aa != '-'\n",
    "        if idx+1 in keep_resis:\n",
    "            if (keep_same_as_consensus and seq_aa in consensus[idx]) or not keep_same_as_consensus:\n",
    "                mapped_seq.append(idx+1)\n",
    "\n",
    "    return mapped_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# a3m output made using https://toolkit.tuebingen.mpg.de/tools/hhblits\n",
    "# output contains insertions (lower case chars), which are cut out during the process_fasta() step\n",
    "\n",
    "name_to_seq = process_fasta(input_a3m)\n",
    "num_aas_by_pos = get_consensus_dict(name_to_seq)\n",
    "perc_aas_by_pos, consensus = get_consensus_dict(name_to_seq)\n",
    "\n",
    "ranked_resis_50 = cutoff_ranked_resis(perc_aas_by_pos, 0.5)\n",
    "ranked_resis_70 = cutoff_ranked_resis(perc_aas_by_pos, 0.7)\n",
    "\n",
    "# update the string below with your protein's name\n",
    "resis_to_keep_50 = renum_ranked_to_seq(input_protein, consensus, name_to_seq, ranked_resis_50, False)\n",
    "resis_to_keep_70 = renum_ranked_to_seq(input_protein, consensus, name_to_seq, ranked_resis_70, False)\n",
    "\n",
    "resis_to_keep_50 += resis_to_lock.copy()\n",
    "resis_to_keep_70 += resis_to_lock.copy()\n",
    "\n",
    "# print outputs for pymol coloring\n",
    "# print('color palegreen, sele and resi ' + '+'.join([str(i) for i in resis_to_keep_70]))\n",
    "# print('color paleyellow, sele and resi ' + '+'.join([str(i) for i in resis_to_keep_50]))\n",
    "\n",
    "# print outputs for MPNN\n",
    "print('70% most frequent residues, unless your aa is different than consensus:')\n",
    "print(f'fixed_positions=\"{\" \".join(str(i) for i in resis_to_keep_70)}\"')\n",
    "print('\\n50% most frequent residues, unless your aa is different than consensus:')\n",
    "print(f'fixed_positions=\"{\" \".join(str(i) for i in resis_to_keep_50)}\"')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
