{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [],
   "source": [
    "# variables to edit!\n",
    "# context = current plasmid with insert\n",
    "# mutation_pos = the nucleotide right before the mutation\n",
    "# addition = the mutation to add in\n",
    "\n",
    "def calc_tm(primer, n):\n",
    "    percent_gc = (primer.count('G') + primer.count('C')) / n\n",
    "    return 81.5 + (41 * percent_gc) - (675 / n)\n",
    "\n",
    "def reverse_primer(p):\n",
    "    p = p.upper()\n",
    "    rev_p = ''\n",
    "    rev_d = {'A': 'T', 'T': 'A', 'G': 'C', 'C': 'G'}\n",
    "    for nt in p:\n",
    "        rev_p += rev_d[nt]\n",
    "    return rev_p[::-1]\n",
    "\n",
    "def count_gc(primer):\n",
    "    return (primer.count('G') + primer.count('C')) / len(primer)\n",
    "\n",
    "def get_fwd_primer_mut(context, mutation_pos, mut):\n",
    "\n",
    "    best_seq = context[mutation_pos-15:mutation_pos] + mut + context[mutation_pos+3:mutation_pos+22]\n",
    "    best_tm = calc_tm(best_seq, len(best_seq))\n",
    "    gc_count = count_gc(best_seq)\n",
    "    if best_tm >= 78. and best_seq[-1] in ('G', 'C') and best_seq[0] in ('G', 'C') and gc_count >= 0.4:\n",
    "        return best_seq\n",
    "\n",
    "    for neg in range(12,16):\n",
    "        for pos in range(18,25):\n",
    "            fwd = context[mutation_pos-neg:mutation_pos] + mut + context[mutation_pos+3:mutation_pos+pos]\n",
    "            tm = calc_tm(fwd, len(fwd))\n",
    "            gc_count = count_gc(fwd)\n",
    "            if tm >= 78. and fwd[-1] in ('G', 'C') and fwd[0] in ('G', 'C') and gc_count >= 0.4:\n",
    "                return fwd\n",
    "            elif tm > best_tm:\n",
    "                best_seq = fwd\n",
    "                best_tm = tm\n",
    "                best_gc = gc_count\n",
    "    \n",
    "    return(f'did not converge with Tm of {best_tm} and %GC of {best_gc}, but best seq is: {best_seq}')\n",
    "\n",
    "def get_rev_primer_mut(context, mutation_pos, mut):\n",
    "\n",
    "    best_seq = reverse_primer(context[mutation_pos-22:mutation_pos] + mut + context[mutation_pos+3:mutation_pos+15])\n",
    "    best_tm = calc_tm(best_seq, len(best_seq))\n",
    "    best_gc = count_gc(best_seq)\n",
    "    if best_tm >= 78. and best_seq[-1] in ('G', 'C') and best_seq[0] in ('G', 'C') and best_gc >= 0.4:\n",
    "        return best_seq\n",
    "\n",
    "    for neg in range(18,25):\n",
    "        for pos in range(12,16):\n",
    "            rev = reverse_primer(context[mutation_pos-neg:mutation_pos] + mut + context[mutation_pos+3:mutation_pos+pos])\n",
    "            tm = calc_tm(rev, len(rev))\n",
    "            gc_count = count_gc(rev)\n",
    "            if tm >= 78. and rev[-1] in ('G', 'C') and rev[0] in ('G', 'C') and gc_count >= 0.4:\n",
    "                return rev\n",
    "            elif tm > best_tm:\n",
    "                best_seq = rev\n",
    "                best_tm = tm\n",
    "                best_gc = gc_count\n",
    "\n",
    "    return(f'did not converge with Tm of {best_tm} and %GC of {best_gc}, but best Tm seq is: {best_seq}')\n",
    "\n",
    "def align_dna(p, context):\n",
    "    results = {k: 0 for k in range(len(context))}\n",
    "\n",
    "    for idx in range(len(context)):\n",
    "        count = 0\n",
    "        for pidx, pnt in enumerate(p):\n",
    "            if idx+pidx < len(context):\n",
    "                if pnt == context[idx+pidx]:\n",
    "                    count += 1\n",
    "        results[idx] = count\n",
    "    \n",
    "    return max(results, key=lambda key: results[key])\n",
    "\n",
    "def align_primers_to_context(fwd_p, rev_p, context, reverse=False):\n",
    "    context = context.upper()\n",
    "\n",
    "    fwd_pos = align_dna(fwd_p.upper(), context)\n",
    "    rev_pos = align_dna(reverse_primer(rev_p.upper()), context)\n",
    "\n",
    "    min_pos = min(fwd_pos, rev_pos)\n",
    "    max_pos = max(fwd_pos+len(fwd_p), rev_pos+len(rev_p))\n",
    "\n",
    "    print(f\"5' {'-' * (fwd_pos-(min_pos-3))}>-{fwd_p}->{'-' * (3 + max_pos - (fwd_pos + len(fwd_p)))} 3' forward primer\")\n",
    "    fwd_match = ''\n",
    "    for idx, nt in enumerate(fwd_p):\n",
    "        if nt == context[idx+fwd_pos]:\n",
    "            fwd_match += ' '\n",
    "        else:\n",
    "            fwd_match += 'X'\n",
    "    print(f\"   {' ' * (fwd_pos-(min_pos-5))}{fwd_match}{' ' * (5 + max_pos - (fwd_pos + len(fwd_p)))}   \")\n",
    "    print(f\"5' {context[min_pos-5:max_pos+5]} 3' context\")\n",
    "    print(f\"3' {reverse_primer(context[min_pos-5:max_pos+5])[::-1]} 5' context\")\n",
    "\n",
    "    # rev_p and context are reversed again ([::-1]) b/c it's going 3' -> 5' in this visualization\n",
    "    rev_match = ''\n",
    "    for idx, nt in enumerate(rev_p[::-1]): \n",
    "        if nt == reverse_primer(context)[::-1][idx+rev_pos]:\n",
    "            rev_match += ' '\n",
    "        else:\n",
    "            rev_match += 'X'\n",
    "    print(f\"   {' ' * (rev_pos-(min_pos-5))}{rev_match}{' ' * (5 + max_pos - (rev_pos + len(rev_p)))}   \")\n",
    "    print(f\"3' {'-' * (rev_pos-(min_pos-3))}<-{rev_p[::-1]}-<{'-' * (3 + max_pos - (rev_pos + len(rev_p)))} 5' reverse primer\")\n",
    "\n",
    "def aa_from_codon(codon):\n",
    "    for k, v in ecoli_codon_dict.items():\n",
    "        if codon in v:\n",
    "            return k\n",
    "\n",
    "ecoli_codon_dict = {\n",
    "    'A': ['GCG', 'GCC', 'GCA', 'GCT'],\n",
    "    'C': ['TGC', 'TGT'],\n",
    "    'D': ['GAT', 'GAC'],\n",
    "    'E': ['GAA', 'GAG'],\n",
    "    'F': ['TTT', 'TTC'],\n",
    "    'G': ['GGC', 'GGT', 'GGG', 'GGA'],\n",
    "    'H': ['CAT', 'CAC'],\n",
    "    'I': ['ATT', 'ATC', 'ATA'],\n",
    "    'K': ['AAA', 'AAG'],\n",
    "    'L': ['CTG', 'TTA', 'TTG', 'CTT', 'CTC', 'CTA'],\n",
    "    'M': ['ATG'],\n",
    "    'N': ['AAC', 'AAT'],\n",
    "    'P': ['CCG', 'CCA', 'CCT', 'CCC'],\n",
    "    'Q': ['CAG', 'CAA'],\n",
    "    'R': ['CGT', 'CGC', 'CGG', 'CGA', 'AGA', 'AGG'], \n",
    "    'S': ['AGC', 'TCT', 'AGT', 'TCC', 'TCA', 'TCG'],\n",
    "    'T': ['ACC', 'ACG', 'ACT', 'ACA'],\n",
    "    'V': ['GTG', 'GTT', 'GTC', 'GTA'],\n",
    "    'W': ['TGG'],\n",
    "    'Y': ['TAT', 'TAC'],\n",
    "    '*': ['TAA', 'TGA', 'TAG']\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CACCGAACACATCACCTTTAACGTATCCTGC\n",
      "GATACGTTAAAGGTGATGTGTTCGGTGTTGCC\n",
      "5' -------->-CACCGAACACATCACCTTTAACGTATCCTGC->--- 3' forward primer\n",
      "                          XXX                       \n",
      "5' ATGTGGGCAACACCGAACACATCTTATTTAACGTATCCTGCCCAGT 3' context\n",
      "3' TACACCCGTTGTGGCTTGTGTAGAATAAATTGCATAGGACGGGTCA 5' context\n",
      "                          XXX                       \n",
      "3' ---<-CCGTTGTGGCTTGTGTAGTGGAAATTGCATAG-<------- 5' reverse primer\n"
     ]
    }
   ],
   "source": [
    "context='DNA_CONTEXT_SEQ_HERE'\n",
    "aa_mut_pos = 1 # int\n",
    "new_aa = 'T' # 1-letter aa code\n",
    "\n",
    "\n",
    "fwd_p = get_fwd_primer_mut(context.upper().strip(), (aa_mut_pos-1)*3, ecoli_codon_dict[new_aa][0])\n",
    "rev_p = get_rev_primer_mut(context.upper().strip(), (aa_mut_pos-1)*3, ecoli_codon_dict[new_aa][0])\n",
    "\n",
    "print(fwd_p)\n",
    "print(rev_p)\n",
    "print('')\n",
    "\n",
    "# check that things look correct \n",
    "align_primers_to_context(fwd_p, rev_p, context)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# batch\n",
    "mutations = [\n",
    "    (1, 'M'),\n",
    "    (2, 'E'), # etc\n",
    "]\n",
    "\n",
    "name_idx = 1\n",
    "for aa_mut_pos, new_aa in mutations:\n",
    "    print(f'mut.{name_idx}')\n",
    "    fwd_p = get_fwd_primer_mut(context.upper().strip(), (aa_mut_pos-1)*3, ecoli_codon_dict[new_aa][0])\n",
    "    rev_p = get_rev_primer_mut(context.upper().strip(), (aa_mut_pos-1)*3, ecoli_codon_dict[new_aa][0])\n",
    "    print(fwd_p)\n",
    "    print(rev_p)\n",
    "    print('')\n",
    "    align_primers_to_context(fwd_p, rev_p, context)\n",
    "    print('')\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "b6928e6c810adfb088ca3848dd9881372d22a791ca390268b33b4d5ab4a28c7f"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
